src/http-gateways/trustless-gateway.md

```diff
@@ -301,6 +302,31 @@ of their presence in the DAG or the value assigned to the "dups" parameter, as
 the raw data is already present in the parent block that links to the identity
 CID.
 
+## CAR `meta` (content type parameter)
+
+The `meta` parameter allows clients to request the server to include additional metadata about the
+CAR to be included at the end of the response body.
+
+This parameter can be used with `version=1` only.
+
+When the parameter is not set, the server must not add any extra CAR blocks to the response.
+
+The metadata block is a regular CAR block with the following properties:
+
+- CID specifies multicodec `car-metada` (0x04ff), see
+  [multicodec#334](https://github.com/multiformats/multicodec/pull/334).
+
+- The payload contains metadata encoded as DAG-CBOR.
+
+The metadata MUST include the following fields:
+
+- `len` - byte length of the CAR data (excluding the metadata block)
+- `b3h` - Blake3 hash (checksum) of the CAR data (excluding the metadata block).
+- `b3h_sig` - A signature over `<len><b3h><request>` using server's Ed2559 identity.
+  - `len` is encoded as `varint`,
+  - `b3h` is encoded as 32 bytes,
+  - The effective query as executed by the gateway. This query is the request url - path and query string arguments.
```

Comment:
I agree that the Spark use case belongs in Userland section. However, the individual keys of the metadata section, and what the client must do to implement them, feels like something that should be in the trustless gateway spec. 

Keys like `car_bytes`, `data_bytes`, `block_count` will be used by Spark but also may be used by others, and the definition of a key must be the same for each use case. E.g. if one client sets `data_bytes` to be the total byte length of blocks and another sets it to be the total byte length of the CAR stream then trustless gateway implementers will need to implement different logic for each use case.

What's more, for the Spark use case, we do not want gateway operators to know that they are serving a Spark request and not some other request. Since the Spark ones will be incentivised and other request may not be, servers can simply provide a good retrieval service to Spark clients and a poor service to other clients.

`car_bytes`, `data_bytes`, `block_count` seem generic enough. The troublesome one is then the Blake3Hash and signature.

Reply:
Perhaps what we need to do is leave this IPIP to concern the metadata block being appended without any constraints on what can be included in it. Then in a separate place, we define a canonical way to include a key value   object in the metadata block and how the server should implement certain useful keys such as `car_bytes` et al

Reply:
I think it would be ok to suggest a key name convention for generic things like  `car_bytes` `data_bytes` and `block_count` in the section that described `meta` parameter, as long it is
- scoped to JSON (perhaps list under explicit `meta=eof[+json]`?)
- change requirement from MUST to SHOULD (convention, not a hard requirement)

Reply:
> I think it would be also ok to have a documented convention for passing a hash of the CAR stream (aka CAR CID) – maybe name it `car_cid` and use CIDv1 with 0x0202 codec – this convention is already used by .storage folks, no need to invent anything new.

+1 to document `car_cid`.

For SPARK, we specifically want a Blake3 hash so that we can use inclusion proofs. That's why we want to use a dedicated field `b3checksum` instead of a more generic `car_cid`.